src/infra/core/tests/tests/test_compact_service_impl.rs

```diff
@@ -0,0 +1,355 @@
+                .push_event(MetadataFactory::set_data_schema().build())
+                .build(),
+        )
+        .await;
+
+    let data_helper = harness.dataset_data_helper(&root_dataset_alias).await;
+
+    let dataset_handle = harness
+        .dataset_repo
+        .resolve_dataset_ref(&root_dataset_alias.as_local_ref())
+        .await
+        .unwrap();
+
+    harness
+        .append_add_data_block(&root_dataset_alias.as_local_ref(), FILE_DATA_ARRAY_SIZE)
+        .await;
+
+    let old_blocks = harness
+        .get_dataset_blocks(&root_dataset_alias.as_local_ref())
+        .await;
+
+    assert_matches!(
+        harness
+            .compact_svc
+            .compact_dataset(
+                &dataset_handle,
+                MAX_SLICE_SIZE,
+                MAX_SLICE_RECORDS,
+                Some(Arc::new(NullCompactionMultiListener {}))
+            )
+            .await,
+        Ok(()),
+    );
+
+    data_helper
+        .assert_last_data_eq(
+            indoc!(
+                r#"
+                message arrow_schema {
+                  REQUIRED INT32 a;
+                  REQUIRED BYTE_ARRAY b (STRING);
+                }
+                "#
+            ),
+            indoc!(
+                r#"
+                +---+---+
+                | a | b |
+                +---+---+
+                | 1 | a |
+                | 2 | b |
+                | 3 | c |
+                | 4 | d |
+                | 5 | e |
+                +---+---+
+                "#
+            ),
+        )
+        .await;
+
+    let new_blocks = harness
+        .get_dataset_blocks(&root_dataset_alias.as_local_ref())
+        .await;
+    assert_eq!(old_blocks.len(), new_blocks.len());
+
+    harness
+        .append_add_data_block(&root_dataset_alias.as_local_ref(), FILE_DATA_ARRAY_SIZE)
+        .await;
+    let old_blocks = harness
+        .get_dataset_blocks(&root_dataset_alias.as_local_ref())
+        .await;
+
+    assert_matches!(
+        harness
+            .compact_svc
+            .compact_dataset(
+                &dataset_handle,
+                MAX_SLICE_SIZE,
+                MAX_SLICE_RECORDS,
+                Some(Arc::new(NullCompactionMultiListener {}))
+            )
+            .await,
+        Ok(()),
+    );
+
+    // 2 Dataslices will be merged in a one slice
+    data_helper
+        .assert_last_data_eq(
+            indoc!(
+                r#"
+            message arrow_schema {
+              REQUIRED INT32 a;
+              REQUIRED BYTE_ARRAY b (STRING);
+            }
+            "#
+            ),
+            indoc!(
+                r#"
+            +---+---+
+            | a | b |
+            +---+---+
+            | 1 | a |
+            | 2 | b |
+            | 3 | c |
+            | 4 | d |
+            | 5 | e |
+            | 1 | a |
+            | 2 | b |
+            | 3 | c |
+            | 4 | d |
+            | 5 | e |
+            +---+---+
```

Comment:
This test doesn't quite tell me that the ordering of blocks is preserved by the merge process - order is hard to tell when two data slices are identical :)

I think we have to write data properly, like in `test_push_ingest.rs` for us to have the `offset` column.

I would also recommend having another test that generates 100+ of small blocks and merges them into one, ensuring the order of `offset`s is preserved.

Let us know if you need help with that.

Reply:
Added new tests and in process found issue with sorting(Fixed as well)